src/commands/deploy.js

```diff
@@ -16,20 +16,263 @@ const SitesCreateCommand = require('./sites/create')
+        )
+      }
+      // Improve the message of permission errors
+      if (e.code === 'EACCES') {
+        error('Permission error when trying to access functions folder')
+      }
+    }
+  }
+
+  if (stat && !stat.isDirectory()) {
+    error('Functions folder must be a path to a directory')
+  }
+
+  return stat
+}
+
+const validateFolders = async ({ deployFolder, functionsFolder, error, log }) => {
+  const deployFolderStat = await validateDeployFolder({ deployFolder, error })
+  const functionsFolderStat = await validateFunctionsFolder({ functionsFolder, error, log })
+  return { deployFolderStat, functionsFolderStat }
+}
+
+const runDeploy = async ({
+  flags,
+  deployToProduction,
+  siteData,
+  api,
+  siteId,
+  deployFolder,
+  configPath,
+  functionsFolder,
+  alias,
+  log,
+  warn,
+  error,
+  exit,
+}) => {
+  let results
+  try {
+    if (deployToProduction) {
+      if (isObject(siteData.published_deploy) && siteData.published_deploy.locked) {
+        log(`\n${NETLIFYDEVERR} Deployments are "locked" for production context of this site\n`)
+        const { unlockChoice } = await inquirer.prompt([
+          {
+            type: 'confirm',
+            name: 'unlockChoice',
+            message: 'Would you like to "unlock" deployments for production context to proceed?',
+            default: false,
+          },
+        ])
+        if (!unlockChoice) exit(0)
+        await api.unlockDeploy({ deploy_id: siteData.published_deploy.id })
+        log(`\n${NETLIFYDEVLOG} "Auto publishing" has been enabled for production context\n`)
+      }
+      log('Deploying to main site URL...')
+    } else {
+      log('Deploying to draft URL...')
+    }
+
+    results = await api.deploy(siteId, deployFolder, {
+      configPath,
+      fnDir: functionsFolder,
+      statusCb: flags.json || flags.silent ? () => {} : deployProgressCb(),
+      draft: !deployToProduction && !alias,
+      message: flags.message,
+      deployTimeout: flags.timeout * 1000 || 1.2e6,
+      syncFileLimit: 100,
+      branch: alias,
+    })
+  } catch (e) {
+    switch (true) {
+      case e.name === 'JSONHTTPError': {
+        warn(`JSONHTTPError: ${e.json.message} ${e.status}`)
+        warn(`\n${JSON.stringify(e, null, '  ')}\n`)
+        error(e)
+        return
+      }
+      case e.name === 'TextHTTPError': {
+        warn(`TextHTTPError: ${e.status}`)
+        warn(`\n${e}\n`)
+        error(e)
+        return
+      }
+      case e.message && e.message.includes('Invalid filename'): {
+        warn(e.message)
+        error(e)
+        return
+      }
+      default: {
+        warn(`\n${JSON.stringify(e, null, '  ')}\n`)
+        error(e)
+        return
+      }
+    }
+  }
+
+  const siteUrl = results.deploy.ssl_url || results.deploy.url
+  const deployUrl = get(results, 'deploy.deploy_ssl_url') || get(results, 'deploy.deploy_url')
+  const logsUrl = `${get(results, 'deploy.admin_url')}/deploys/${get(results, 'deploy.id')}`
+
+  return {
```

Comment:
This information was used here https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L241 and here https://github.com/netlify/cli/blob/39da8081eaca6e2c8ba58aa782541199e0df48fc/src/commands/deploy.js#L263